More lwg 22 3 11 (#210)
ORNL/cpp-proposals-pub: P0009/P0009.md

<b>22.7.� Overview [mdspan.terms]</b>

[1]{.pnum} A _multidimensional index space_ is a Cartesian product of integer intervals.
Each interval can be represented by a half-open range $[L_i, U_i)$, where $L_i$ and $U_i$ are the lower and upper bounds of the $i^{th}$ dimension.
The _rank_ of a multidimensional index space is the number of intervals it represents.
The _size_ of a multidimensional index space is the product of $U_i-L_i$ for each dimension $i$ if its rank is greater than $0$, and $1$ otherwise.

> In the following representations of a multidimensional index space, such as `extents`, will be used synonymously with the index space itself.

Reviewer: ```suggestion
In what follows, any of various representations of a multidimensional index space, such as `extents`, will be used synonymously with the index space itself.
```

Author: I moved that sentence to extents intro itself.

> In the following representations of a multidimensional index space, such as `extents`, will be used synonymously with the index space itself.

Author: I moved this to extents. What do you think?

   * [1.2]{.pnum} `m` denotes a (possibly const) value of type `M`.

>    * [1.3]{.pnum} `i` and `j` are packs of (possibly const) integers which are multidimensional indices into `m.extents()` ([mdspan.terms]).

Reviewer: Is the idea that `i` and `j` are in the multidimensional index space represented by `m.extents()`?  If so, then would you consider changing "in" to "into"?
```suggestion
   * [1.3]{.pnum} `i` and `j` are packs of (possibly const) integers which are multidimensional indices in `m.extents()` ([mdspan.terms]).
```

Author: ok

typename M::layout_type
```
[4]{.pnum} *Result:* A type `MP` which meets the layout mapping policy requirements ([mdspan.layoutpolicy.reqmts]) and for which 
                     `typename MP::template mapping<typename M::extents_type>` denotes the type `M`.

```c++
m.extents()
```
[5]{.pnum} *Result:* `const typename M::extents_type&`

```c++
m(i...)
```
[6]{.pnum} *Result:* `typename M::size_type`

> [7]{.pnum} *Returns:* A value less then `numeric_limits<typename M::size_type>::max - 1`.

Reviewer: ```suggestion
[7]{.pnum} *Returns:* A value less then `numeric_limits<typename M::size_type>::max() - 1`.
```

```c++
m(i...) == m(static_cast<typename M::size_type>(i)...)
```

[8]{.pnum} *Result:* `bool`

[9]{.pnum} *Value:* `true`

```c++
m.required_span_size()
```
[10]{.pnum} *Result:* `typename M::size_type`

[11]{.pnum} *Returns:* If the multidimensional index space `m.extents()` has size $0$, then `0`,

>       else `1` plus the maximum value of `m(i...)` for all `i`.

Reviewer: LWG really wanted us not to specify "for all `i` in the domain"?  This sounds uncomfortably like "for all integer packs `i`."

Author: but i is defined to be in the domain

```c++
m.is_unique()
```
[12]{.pnum} *Result:* `bool`

[13]{.pnum} *Returns:* `true` only if for every `i` and `j` where `(i != j || ...)` is `true`, `m(i...) != m(j...)` is `true`.

>      <i>[Note:</i> A mapping may return `false` even if the above condition is met.

Reviewer: Excellent note; thank you!

                   For certain layouts it may not be feasible to determine efficiently whether the layout is unique.<i>— end note]</i>

```c++
m.is_contiguous()
```
[14]{.pnum} *Result:* `bool`

[15]{.pnum} *Returns:* `true` only if for all $k$ in the range $[0,$ `m.required_span_size()` $)$ there exists an `i` such that `m(i...)` equals $k$.
     <i>[Note:</i> A mapping may return `false` even if the above condition is met.
                   For certain layouts it may not be feasible to determine efficiently whether the layout is contiguous.<i>— end note]</i>

```c++
m.is_strided()
```
[16]{.pnum} *Result:* `bool`

[17]{.pnum} *Returns:* `true` only if for every rank index $r$ of `m.extents()` there exists an integer $s_r$

>        such that, for all `i` where `(i+`$d_r$`)` is a multidimensional index into `m.extents()` ([mdspan.terms]),

Reviewer: Please see above note on "in" vs. "into."
